HDF5 Logging

b-butler opened the pull request:
## Description

Adds a HDF5 logger backend for use in HOOMD. We use HDF5 resizable datasets with specified chunk sizes for fast read performance for typical workflows.
<!-- Describe your changes in detail. -->

## Motivation and context

This provides a small performant binary file object for logger backends.
<!--- Why is this change required? What problem does it solve? -->

<!-- Replace ??? with the issue number that this pull request resolves. -->
Resolves #???

## How has this been tested?

Tests have been added, documentation written, and doc-tests made.
<!--- Please describe in detail how you tested your changes. -->

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Change log

<!-- Propose a change log entry. -->
```
Added: HDF5 logger backend (hoomd.write.HDF5Logger)
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

b-butler committed:
feat: Add HDF5 logger.

b-butler committed:
feat (WIP): Add working and efficient HDF5 writer.

Code has been tested on a ~10,000 particle simulation with logging every
frame.

b-butler committed:
feat: Switch to use interal custom actions for HDF5 logger.

Also document class.

b-butler committed:
feat: Add flush method to HDF5Logger.

b-butler committed:
test: Add tests for the HDF5 logger backend.

b-butler committed:
refactor: Only import HDF5Logger if h5py is available

b-butler committed:
doc: Add HDF5Logger to documentation.

b-butler committed:
doc: Add a tested example to HDF5Logger.

b-butler committed:
fix: HDF5Logger errors and test errors.

b-butler committed:
fix: hoomd.logging.LoggerCategories._get_string_list

Removes NONE from the list.

b-butler committed:
feat: Make HDF5Logger objects picklable.

b-butler committed:
doc: Add h5py to documentation requirements.

b-butler committed:
fix: Documentation test example

b-butler committed:
feat: Add an accepted category class attr for HDF5Logger

b-butler committed:
test: Fix pytest module skip syntax.

tommy-waltmann commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,225 @@
+        filename (str): The filename of the HDF5 file to write to.
+        logger (hoomd.logging.Logger): The logger instance to use for querying
+            log data.
+        mode (`str`, optional): The mode to open the file in. Defaults to "a".
```
Let's add more documentation on supported modes. If they mirror h5 file modes exactly, let's add a link to h5's documentation on that.

tommy-waltmann commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,225 @@
+        mode (`str`, optional): The mode to open the file in. Defaults to "a".
+
+    Attributes:
+        cls.accepted_categories (hoomd.logging.LoggerCategories): The enum value
+            for all accepted categories for `HDF5Logger` instances.
```
Can we list out the accepted categories somewhere, or would that list be cumbersomely long?

b-butler replied:
That would be quite long (there are only three disallowed categories), but we can mention that.

tommy-waltmann commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,225 @@
+        """
+        self._fh.flush()
+
+    def _create_database(self, key: str, shape, dtype, chunk_size):
```
```suggestion
    def _create_dataset(self, key: str, shape, dtype, chunk_size):
```
Does this seem like a more appropriate name, given the name of the h5 call in its implementation and context it gets used in?

b-butler replied:
Good point.

tommy-waltmann commented on hoomd/write/__init__.py:
Should we add some description of the HDF5 writer to the docstring at the top of this file? Upon further inspection, it seems we forgot to do this with the burst writer in a previous PR.

tommy-waltmann commented on hoomd/md/pytest/test_hdf5.py:
```diff
@@ -0,0 +1,128 @@
+    return sim
+
+
+@pytest.fixture
+def logger():
+    logger = hoomd.logging.Logger(categories=['scalar'])
+    return logger
```
Does a very simple one-liner like this warrant a fixture?

b-butler replied:
Not really, just a stylistic preference, but I will add the line to the tests.

tommy-waltmann reviewed (changes requested):
Great work on this so far, I just have a few comments related to code style and documentation.

b-butler committed:
fix: calculation of accepted_categories in HDF5Logger.

b-butler committed:
feat: Allow logger to be set with LoggerCategories instance.

b-butler committed:
test: Test setting Logger.categories to LoggerCategories

b-butler committed:
doc: Add Burst and HDF5Logger to writer main documentation

b-butler committed:
refactor: HDF5Logger _create_database -> _create_dataset

b-butler committed:
doc: Improve documentation for HDF5Logger

b-butler committed:
test: Remove test_hdf5.py:logger fixture

b-butler committed:
doc: Fix documentation builds

b-butler committed:
doc: Add h5py to full testing requirements.

b-butler committed:
doc: Actual documentation fix.

I think. I have tested locally every time without error, so we will see
_/(-_-)\_.

b-butler committed:
fix: MPI behavior for HDF5Logger.

b-butler commented:
@joaander we in gloterlab/ci-containers

https://github.com/glotzerlab/ci-containers/blob/5795b26faaab6b64d48abae78dff4b629097069e/ubuntu22.04/py310/Dockerfile#L28

exclude h5py from installation in the docker file. Can we undo that, or is there a reason for the commenting out?

b-butler committed:
fix: HDF5Logger MPI behavior.

Use communicator upon attaching for opening the file only on the root
rank.

b-butler committed:
refactor: Use helper class to skip file handle actions

for HDF5Logger.

b-butler committed:
test: Fix HDF5Logger MPI test function errors.

joaander commented on hoomd/write/__init__.py:
```diff
@@ -30,3 +33,7 @@
 from hoomd.write.gsd_burst import Burst
 from hoomd.write.dcd import DCD
 from hoomd.write.table import Table
+try:
```
Move the try/except around `import h5py` in `write/hdf5.py`. Will also need to add an error check (e.g. if `h5py is None`) check in `HDF5Logger`. This way, HOOMD developers that modify `write/hd5.py` will see the cause of any `ImportError` related to `hdf5.py` and the `HDF5Logger` class is available (e.g. for sphinx doc builds) when `h5py` is not installed.

joaander commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,272 @@
+            if category == "scalar":
+                data_shape = (1,)
+                dtype = "f8" if isinstance(value, float) else "i8"
+                chunk_size = (500,)
```
```suggestion
                chunk_size = (512,)
```
Not really necessary, but this will make the chunks nicely map neatly into power of 2 filesystem block sizes.

joaander commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,272 @@
+                    value = np.asarray(value)
+                data_shape = (1,) + value.shape
+                dtype = value.dtype
+                chunk_size = [dim if dim <= 10 else dim for dim in data_shape]
+                chunk_size[0] = 1
```
```suggestion
                chunk_size = data_shape
```
Unless this code is an error and you intend to perform some modification to the chunk size in the list comprehension.

Also, I would assume that `chunk_size = data_shape` works very well for typical per-particle, per-bond, and other large arrays. It might not work so well for fixed length 2 arrays. For parity with the scalar chunk size above, try:
```
chunk_size[0] = max(4096 // value.nbytes, 1)
```
If this works well, remove magic numbers by making 4096 a named constant and using it in both places we compute a chunk size in dimensions.

b-butler replied:
I originally did, but it did not change performance. Good catch.

joaander commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,272 @@
+            log data.
+        mode (`str`, optional): The mode to open the file in. Available values
+            are "w", "x" and "w-", "a", and "r+". Defaults to "a". See the
+            h5py_ documentation for more).
```
```suggestion
            h5py_ documentation for more details).
```

joaander commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,272 @@
+    .. code-block:: python
+
+        logger = hoomd.logging.Logger(
+            hoomd.write.HDF5Logger.accepted_categories)
```
Nice!

joaander commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,272 @@
+
+        logger = hoomd.logging.Logger(
+            hoomd.write.HDF5Logger.accepted_categories)
+        h5_writer = hoomd.write.HDF5Logger(10_000, "simulation-log.h5", logger)
```
```suggestion
        h5_writer = hoomd.write.HDF5Logger(
            trigger=10_000,
            filename=h5_filename,
            logger=logger)
```
I like to be explicit with keyword arguments in examples. It makes the code more readable.

Also, per our discussions on #1586, create a `h5_filename` variable in the invisible codeblock to use here:
```
h5_filename = tmp_path / 'file.h5'
```
Without the tmp_path fixture, the file will be placed in the current working directory when executing `pytest`.

joaander commented on sphinx-doc/requirements.txt:
```diff
@@ -1,3 +1,4 @@
+h5py
```
```suggestion
```
With my suggested change to the `try / except ImportError` above, this is no longer needed.

joaander commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,272 @@
+
+        Writes out a frame from the composed logger.
+        """
+        self._action.act()
```
How does this work? `act` takes a `timestep` argument. Also, should we be providing a public facing `write` API at all?

b-butler replied:
The timestep is optional in many custom writers. Also, we provide a `write` method in `_InternalCustomWriter`. We could remove that (can't remember why I put it there), but it would technically be a breaking change.

joaander commented on sphinx-doc/module-hoomd-write.rst:
```diff
@@ -24,5 +25,8 @@ hoomd.write
     :show-inheritance:
 
+    .. autoclass:: HDF5Logger(trigger, filename, logger, mode="a")
+        :members:
```
```suggestion
        :members: flush write
```
`flush` currently doesn't appear in the documentation. If I recall correctly, the internal action's methods are somehow automatically made available to the parent - right? Does this work to make flush appear in the sphinx docs? Or do we need to define a placeholder in `HDF5Logger`?

joaander reviewed (changes requested):
Thanks! This will be very helpful for space-efficient log-only files. I have some suggestions for improvement.

h5py for py310 is available in the 2023.07.20 ci container build. At the time I initially created the py310 container, wheels were not available for h5py. This is no longer the case.

The failing test demonstrates that h5py is not actually an optional test dependency as documented. We should fix that.

b-butler committed:
refactor: Move h5py try except to _HDF5LoggerInternal

b-butler committed:
refactor: use class constants over magic numbers HDF5Logger

Use class constants when determining chunk sizes.

b-butler committed:
doc: Improve HDF5Logger documentation.

b-butler committed:
fix: Add timestep to HDF5Logger method.

b-butler committed:
doc: Clean up wording.

b-butler committed:
Merge branch 'trunk-minor' into feat/h5-logger

b-butler committed:
test: Fix doc test for Logger.__setitem__

b-butler committed:
doc: Fix directive spacing

b-butler committed:
test: make hdf5.py doctests condition on h5py.

tommy-waltmann commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,295 @@
+        This class cannot handle string, strings, or object loggables.
+
+    Args:
+        trigger (hoomd.trigger.trigger_like): The trigger to determine when to
+            run the HDF5 backend.
```
```suggestion
        trigger (hoomd.trigger.trigger_like): The trigger to determine when to
            write to the HDF5 file.
```

joaander replied:
Also fix the description in `Attributes:` below.

tommy-waltmann commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,295 @@
+        self.__dict__ = state
+
+
+class HDF5Logger(_InternalCustomWriter):
```
Would this class more appropriately be named simply `HDF5` instead of `HDF5Logger`? It only writes log data (i.e. not snapshots), but its still a writer and hoomd's other writers are named according to the type of file they write to.

If we change this class's name, we may also want to change the name of the internal class.

joaander replied:
`HDF5Logger` is awkward because it is _not_ a `Logger`. At the same time putting `HDF5` alongside `GSD` and `DCD` gives some hint that it might also write the system state when it does not.

What do you think of `HDF5Log` as a compromise solution?

tommy-waltmann replied:
I would be okay with that. My idea to fix the potential user confusion was to clarify repeatedly in the class documentation that it writes log data only, which I've tried to do to some extent in the other suggestions in my PR review.

tommy-waltmann commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,295 @@
+
+
+class HDF5Logger(_InternalCustomWriter):
+    """An HDF5 logger backend.
```
```suggestion
    """Write loggable simulation data to HDF5 files.
```

tommy-waltmann commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,295 @@
+class HDF5Logger(_InternalCustomWriter):
+    """An HDF5 logger backend.
+
+    This class handles scalar and array data storing them in HDF5 resizable
+    datasets.
```
```suggestion
    This class stores resizable scalar and array data in the HDF5 file format.
```

tommy-waltmann reviewed (changes requested):
This looks to be in really good shape, I just have some naming and re-wording suggestions to clarify the documentation.

joaander commented on hoomd/logging.py:
```diff
@@ -625,25 +628,49 @@ class they come from. For instance, the `hoomd.md.pair.LJ` class has a
+    .. rubric:: Examples
+
+    There are various ways to create a logger with different available
+    loggables. To allow all loggable categories just create a logger.
```
```suggestion
    loggables. Create a `Logger` with no options to allow all categories.
```
"just" belittles the reader.

joaander commented on hoomd/logging.py:
```diff
@@ -625,25 +628,49 @@ class they come from. For instance, the `hoomd.md.pair.LJ` class has a
 
     .. code-block:: python
 
+        from hoomd.logging import LoggerCategories
+
         logger = hoomd.logging.Logger()
```
```suggestion
        logger = hoomd.logging.Logger()
```
Unnecessary import.

joaander commented on hoomd/logging.py:
```diff
@@ -625,25 +628,49 @@ class they come from. For instance, the `hoomd.md.pair.LJ` class has a
+
         logger = hoomd.logging.Logger()
+
+    When restricting to a subset of categories you can use a list of strings.
+
+    .. code-block:: python
+
+        logger = hoomd.logging.Logger(categories=["string", "strings"])
+
+    The `LoggerCategories` enum values can also be used directly.
+
+    .. code-block:: python
+
+        logger = hoomd.logging.Logger(
+            categories=LoggerCategories.particle | LoggerCategories.bond)
```
```suggestion
    Use a list of strings to log a subset of categories:

    .. code-block:: python

        logger = hoomd.logging.Logger(categories=["string", "strings"])
```
I think using the enum directly is an advanced topic and therefore doesn't need an example. Users that need it can infer the appropriate syntax from the reference documentation.

joaander commented on hoomd/md/pytest/test_hdf5.py:
```diff
@@ -0,0 +1,128 @@
+import hoomd
+import hoomd.write
+
+try:
+    import h5py
+except ImportError:
+    pytestmark = pytest.skip("h5py required to test this feature.",
+                             allow_module_level=True)
```
```suggestion
pytest.importorskip("h5py")
```

joaander commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,295 @@
+        logger = hoomd.logging.Logger(
+            hoomd.write.HDF5Logger.accepted_categories)
+        h5_writer = hoomd.write.HDF5Logger(
+            trigger=10_000,
```
```suggestion
            trigger=hoomd.trigger.Periodic(10_000),
```
Explicit is better than implicit.

joaander commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,295 @@
+
+        Without calling this, data may be stored in the h5py.File object without
+        being written to disk yet.
+        """
```
```suggestion
        .. rubric:: Examples
        
        Flush one writer:
        
        .. skip: if(h5py_not_available)
        
        .. code-block:: python
        
            hdf5_log_writer.flush()
            
        Flush all write buffers:
        
        .. skip: if(h5py_not_available)
        
        .. code-block:: python
        
            for writer in simulation.operations.writers:
                if hasattr(writer, 'flush'):
                    writer.flush()
        """
```
Will need to move the invisible setup code up higher in the document for this to pass tests.

joaander commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,295 @@
+
+
+        simulation = hoomd.util.make_example_simulation()
+        h5_filename = tmp_path / "simulation_log.h5"
```
```suggestion
        hdf5_filename = tmp_path / "simulation_log.h5"
```
Use the same name throughout. We don't abbreviate with h5 anywhere else. Alternately, use H5 everywhere: e.g. name the class `H5Log` to be consistent with the file extension.

joaander commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,295 @@
+
+        logger = hoomd.logging.Logger(
+            hoomd.write.HDF5Logger.accepted_categories)
+        h5_writer = hoomd.write.HDF5Logger(
```
```suggestion
        hdf5_log_writer = hoomd.write.HDF5Logger(
```
Match the (expected new) class name.

joaander commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,295 @@
+            hoomd.write.HDF5Logger.accepted_categories)
+        h5_writer = hoomd.write.HDF5Logger(
+            trigger=10_000,
+            filename=h5_filename,
+            logger=logger)
+        simulation.operations += h5_writer
```
```suggestion
            filename=hdf5_filename,
            logger=logger)
        simulation.operations += hdf5_writer
```

joaander commented on hoomd/write/hdf5.py:
```diff
@@ -0,0 +1,295 @@
+        """Write out data to the HDF5 file.
+
+        Writes out a frame from the composed logger.
+        """
```
If I understand correctly, users can call `write` to add frames to the log that are not triggered by `trigger`.
We need to:
1) Add an example snippet.
2) Test `write` in the unit test.
and 3) Explain clearly that calling `write` will add an additional frame even if not triggered.

One issue with `write` is that the fields requiring any of the the flags `ROTATIONAL_KINETIC_ENERGY`, `PRESSURE_TENSOR`, or `EXTERNAL_FIELD_VIRIAL` may be NaN. Those quantities may or may not be available depending on whether the last timestep executed had those flags set by some other means.

joaander reviewed (changes requested):
Thanks! I noted some more consistency and documentation updates to improve this further.

Looking at the implementation of `write` (which accepts `timestep=None`), I realize that this writer doesn't output the timestep.  As a user, I would want to know what timestep my log frames are on, especially in cases with non-periodic triggers.

`GSD` writes the timestep by default. Should we do that here? Or nudge users into using the `Simulation.timestep` loggable by including that in the example?

b-butler committed:
doc: Improve HDF5Logger wording

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

b-butler committed:
refactor: Apply suggested changes to HDF5Logger code.

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>
Co-authored-by: Tommy Waltmann <53307607+tommy-waltmann@users.noreply.github.com>

b-butler committed:
refactor: Rename HDF5Logger to HDF5Log

b-butler committed:
doc: Add HDF5Log.flush example

b-butler committed:
doc: Add example for HDF5Log.write.

b-butler committed:
doc: Improve HDF5Log.write wording.

b-butler committed:
test: Add test for HDF5Log.write.

b-butler committed:
test: Rename h5_writer instances to hdf5_writer

b-butler committed:
test: Fix doctest for HDF5Log

b-butler committed:
test: Fix skipping of HDF5Log doctests.

Tests were not skipping properly in the absence of h5py.

b-butler committed:
doc: Fix sphinx syntax for HDF5Log.

joaander reviewed (approved):
Thanks!

tommy-waltmann reviewed (approved):
LGTM
